/***************************************************************************
 *   graphselectdlg.cpp                                                    *
 *   Copyright (C) 2014 by Daniel Suson                                    *
 *   suson@calumet.purdue.edu                                              *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 *                                                                         *
 *   In addition, as a special exception, the copyright holders give       *
 *   permission to link the code of this program with any edition of       *
 *   the Qt library by Trolltech AS, Norway (or with modified versions     *
 *   of Qt that use the same license as Qt), and distribute linked         *
 *   combinations including the two.  You must obey the GNU General        *
 *   Public License in all respects for all of the code used other than    *
 *   Qt.  If you modify this file, you may extend this exception to        *
 *   your version of the file, but you are not obligated to do so.  If     *
 *   you do not wish to do so, delete this exception statement from        *
 *   your version.                                                         *
 ***************************************************************************
 * Description:                                                            *
 *                                                                         *
 * This class defines the graph select dialog structure associated with    *
 * the Qt4 user interface.                                                 *
 * ----------------------- Change Log ------------------------------------ *
 * Date    Name         Description                                        *
 *                                                                         *
 ***************************************************************************
 * This software evolved from work that was done by Eric Hivon at the      *
 * California Institute of Technology on the MASTER algorithm (Hivon, E.,  *
 * et al., ApJ, vol 567, p 2).  It adapts or interfaces with these         *
 * packages:                                                               *
 *                                                                         *
 * HealPix         (http://www.eso.org/science/healpix/)                   *
 * FTOOLS & FITSIO (http://heasarc.gsfc.nasa.gov/docs/corp/software.html)  *
 *                                                                         *
 * Any publications or presentations that use results generated by this    *
 * program must reference the fact that it was developed by Daniel Suson   *
 ***************************************************************************/
#include "graphselectdlg.h"
#include "ui_graphselectdlg.h"

graphSelectDialog::graphSelectDialog() :
                   ui(new Ui::graphSelectDialog) {
  /* set up the user interface first */
  ui->setupUi(this);
  
  /* initialize internal variables*/
  graphType = associatedSpectrum::Null;
  dirty = false;
  activeGraphs = 0;
  
  /* enable all graphs */
  ui->noiseButton->setEnabled(true);
  ui->filterButton->setEnabled(true);
  ui->beamButton->setEnabled(true);
  ui->dataButton->setEnabled(true);
  ui->weightButton->setEnabled(true);
  ui->combinedDataButton->setEnabled(true);
  ui->spectDataButton->setEnabled(true);
  ui->ensembleButton->setEnabled(true);
   
  /* set up signals and slots */
  connect(ui->buttonBox, &QDialogButtonBox::accepted, this, &graphSelectDialog::finalize);
  connect(ui->buttonBox, &QDialogButtonBox::rejected, this, &graphSelectDialog::cancel);
  connect(ui->buttonBox, &QDialogButtonBox::helpRequested, this, &graphSelectDialog::help);
}

graphSelectDialog::~graphSelectDialog() {
  delete ui;
}

void graphSelectDialog::validate() {
  dirty = false; // assume nothing has changed
  
  ASSOCIATEDSPECTRUM oldGraphType = graphType;
  if (ui->noiseButton->isChecked())
    graphType = associatedSpectrum::TransformedNoiseSpectrum;
  if (ui->filterButton->isChecked())
    graphType = associatedSpectrum::TransformedFilterSpectrum;
  if (ui->beamButton->isChecked())
    graphType = associatedSpectrum::TransformedBeamSpectrum;
  if (ui->dataButton->isChecked())
    graphType = associatedSpectrum::TransformedDataSpectrum;
  if (ui->weightButton->isChecked())
    graphType = associatedSpectrum::TransformedWeightsSpectrum;
  if (ui->combinedDataButton->isChecked())
    graphType = associatedSpectrum::WeightedTransformSpectrum;
  if (ui->spectDataButton->isChecked())
    graphType = associatedSpectrum::SpectralDataSpectrum;
  if (ui->ensembleButton->isChecked())
    graphType = associatedSpectrum::EnsembleDataSpectrum;

  if (graphType != oldGraphType)
    dirty = true;
  
  return;
}

void graphSelectDialog::configure(unsigned int availableGraphs) {
  dirty = false;
  activeGraphs = availableGraphs;
  
  ui->noiseButton->setEnabled(false);
  ui->filterButton->setEnabled(false);
  ui->beamButton->setEnabled(false);
  ui->dataButton->setEnabled(false);
  ui->weightButton->setEnabled(false);
  ui->combinedDataButton->setEnabled(false);
  ui->spectDataButton->setEnabled(false);
  ui->ensembleButton->setEnabled(false);

  ASSOCIATEDSPECTRUM graphType = associatedSpectrum::Null;
  int type = static_cast<int>(graphType);
  
  while (true) { // run over graphable data types
    if(graphType == associatedSpectrum::Null) {
      graphType = static_cast<ASSOCIATEDSPECTRUM>(++type);
      continue;
    }
    
    if (graphType == associatedSpectrum::ASSOCIATEDSPECTRUM_LIMIT)
      break;
    
    bool selected = activeGraphs & (1 << (type-1));
    switch (graphType)
    {
      case associatedSpectrum::TransformedNoiseSpectrum:
        if (selected)
          ui->noiseButton->setEnabled(true);
        break;
      case associatedSpectrum::TransformedFilterSpectrum:
        if (selected)
          ui->filterButton->setEnabled(true);
        break;
      case associatedSpectrum::TransformedBeamSpectrum:
        if (selected)
          ui->beamButton->setEnabled(true);
        break;
      case associatedSpectrum::TransformedDataSpectrum:
        if (selected)
          ui->dataButton->setEnabled(true);
        break;
      case associatedSpectrum::TransformedWeightsSpectrum:
        if (selected)
          ui->weightButton->setEnabled(true);
        break;
      case associatedSpectrum::WeightedTransformSpectrum:
        if (selected)
          ui->combinedDataButton->setEnabled(true);
        break;
      case associatedSpectrum::SpectralDataSpectrum:
        if (selected)
          ui->spectDataButton->setEnabled(true);
        break;
      case associatedSpectrum::EnsembleDataSpectrum:
        if (selected)
          ui->ensembleButton->setEnabled(true);
        break;
    }
    graphType = static_cast<ASSOCIATEDSPECTRUM>(++type);
  }

  exec();
}

void graphSelectDialog::reset() {
  activeGraphs = 0;
  graphType = associatedSpectrum::Null;
  ui->noiseButton->setEnabled(true);
  ui->filterButton->setEnabled(true);
  ui->beamButton->setEnabled(true);
  ui->dataButton->setEnabled(true);
  ui->weightButton->setEnabled(true);
  ui->combinedDataButton->setEnabled(true);
  ui->spectDataButton->setEnabled(true);
  ui->ensembleButton->setEnabled(true);
  dirty = false;
}

void graphSelectDialog::finalize() {
  validate();
  Q_EMIT graphSelected(graphType);
  accept();
}

void graphSelectDialog::cancel() {
  if (dirty)
    dirty = false;
  close();
}

void graphSelectDialog::help() {

}
